src/rdkafka_telemetry.c

```diff
@@ -0,0 +1,679 @@
+                    rk->rk_telemetry.push_interval_ms * 1000,
+                    rd_kafka_telemetry_fsm_tmr_cb, (void *)rk);
+        } else { /* error */
+                rd_kafka_dbg(rk, TELEMETRY, "PUSHERR",
+                             "PushTelemetryRequest failed: %s",
+                             rd_kafka_err2str(err));
+                // Non-retriable errors
+                if (err == RD_KAFKA_RESP_ERR_INVALID_REQUEST ||
+                    err == RD_KAFKA_RESP_ERR_INVALID_RECORD) {
+                        rd_kafka_log(
+                            rk, LOG_WARNING, "TELEMETRY",
+                            "PushTelemetryRequest failed with non-retriable "
+                            "error: %s. Stopping telemetry.",
+                            rd_kafka_err2str(err));
+                        rd_kafka_telemetry_set_terminated(rk);
+                        return;
+                }
+
+                if (err == RD_KAFKA_RESP_ERR_TELEMETRY_TOO_LARGE) {
+                        rd_kafka_log(
+                            rk, LOG_WARNING, "TELEMETRY",
+                            "PushTelemetryRequest failed because of payload "
+                            "size too large: %s. Continuing telemetry.",
+                            rd_kafka_err2str(err));
+                        rk->rk_telemetry.state =
+                            RD_KAFKA_TELEMETRY_PUSH_SCHEDULED;
+                        rd_kafka_timer_start_oneshot(
+                            &rk->rk_timers, &rk->rk_telemetry.request_timer,
+                            rd_false, rk->rk_telemetry.push_interval_ms * 1000,
+                            rd_kafka_telemetry_fsm_tmr_cb, (void *)rk);
+                        return;
+                }
+
+                rd_ts_t next_scheduled =
+                    err == RD_KAFKA_RESP_ERR_UNKNOWN_SUBSCRIPTION_ID
+                        ? 0
+                        : rk->rk_telemetry.push_interval_ms * 1000;
+
+                rk->rk_telemetry.state =
+                    RD_KAFKA_TELEMETRY_GET_SUBSCRIPTIONS_SCHEDULED;
+                rd_kafka_timer_start_oneshot(
+                    &rk->rk_timers, &rk->rk_telemetry.request_timer, rd_false,
+                    next_scheduled, rd_kafka_telemetry_fsm_tmr_cb, (void *)rk);
+        }
+}
+
+/**
+ * @brief This method starts the termination for telemetry and awaits
+ * completion.
+ *
+ * @locks none
+ * @locks_acquired rk_telemetry.lock
+ * @locality app thread (normal case) or the main thread (when terminated
+ *           during creation).
+ */
+void rd_kafka_telemetry_await_termination(rd_kafka_t *rk) {
+        rd_kafka_op_t *rko;
+
+        /* In the case where we have a termination during creation, we can't
+         * send any telemetry. */
+        if (thrd_is_current(rk->rk_thread) ||
+            !rk->rk_conf.enable_metrics_push) {
+                /* We can change state since we're on the main thread. */
+                rk->rk_telemetry.state = RD_KAFKA_TELEMETRY_TERMINATED;
```

Comment:
Here it can call `rd_kafka_telemetry_set_terminated(rk);` to avoid repeating that part of the code.

Reply:
Added a call to `rd_kafka_telemetry_set_terminated(rk);` before return